asm-utils/src/reachability.rs

```diff
@@ -10,24 +10,34 @@ use crate::ast::{Argument, Expression, FunctionOpKind, Register, Statement};
 /// Processes the statements and removes all statements and objects that are
 /// not reachable from the label `label`.
 /// Keeps the order of the statements.
-pub fn filter_reachable_from<R: Register, F: FunctionOpKind, A: Architecture>(
+pub fn filter_reachable_from<'a, R: Register, F: FunctionOpKind, A: Architecture>(
     label: &str,
     statements: &mut Vec<Statement<R, F>>,
-    objects: &mut BTreeMap<String, Vec<DataValue>>,
-) {
+    data_sections: &'a mut Vec<Vec<(Option<String>, Vec<DataValue>)>>,
+) -> HashSet<&'a str> {
     let replacements = extract_replacements(statements);
     let replacement_refs = replacements
         .iter()
         .map(|(k, v)| (k.as_str(), v.as_str()))
         .collect();
-    let referenced_labels =
-        find_reachable_labels::<R, F, A>(label, statements, objects, &replacement_refs)
-            .into_iter()
-            .map(|s| s.to_owned())
-            .collect::<HashSet<_>>();
+    let (referenced_labels, referenced_data_sections) =
+        find_reachable_labels::<R, F, A>(label, statements, data_sections, &replacement_refs);
+
+    {
+        let mut idx = 0usize;
+        data_sections.retain(|_| {
+            let must_retain = referenced_data_sections.contains(&idx);
```

Comment:
What's this index? Why is it growing?

Reply:
`data_sections` is a Vec. This is the Vec index of the current iteration of the `retain()` loop.